man/fstrm_dump.1
.TH fstrm_dump 1

.SH NAME

fstrm_dump \- Display metadata and contents of Frame Streams file.

.SH SYNOPSIS

.B fstrm_dump \fIinput-file\fB [\fIoutput-file\fB]

.SH DESCRIPTION

.B fstrm_dump
opens
.I input-file
and prints its framing metadata to \fIstderr\fR and frame data to \fIstdout\fR.
The
.I input-file
may be "-" to read Frame Streams data from standard input.

Frame data is printed as a single line quoted string with non-printable
characters replaced by backslash-prefixed hex escape sequences. For example,
a frame containing "Hello, world\\n" would have its data printed as:

	 "Hello, world\\x0a"

The only framing metadata expected in a Frame Streams file are the
start frame with a content type field, the data frame lengths, and
the stop frame. These are printed, respectively, as:

	FSTRM_CONTROL_START

	FSTRM_CONTROL_FIELD_CONTENT_TYPE (\fIN\fR bytes)
.br
	 "\fIcontent-type\fR"

	Data frame (\fIN\fR) bytes

	FSTRM_CONTROL_STOP

If
.B fstrm_dump
is given the second
.I output-file
parameter, input frames are additionally written to
.I output-file.
This is mainly useful for regression testing of the
.I fstrm_file_writer
code.

.SH SEE ALSO

.BR fstrm_capture (1),
.BR fstrm_replay (1),
.br
Frame Streams C Library \fBhttps://farsightsec.github.io/fstrm\fR


ChangeLog
fstrm (0.4.0)

  * Added manual pages for fstrm_capture and fstrm_dump.

  * Added new tool, fstrm_replay, for replaying saved Frame Streams data
    to a socket connection.

  * Adds TCP support. Add tcp_writer to the core library which
    implements a bi-directional Frame Streams writer as a TCP socket
    client. Introduces new developer API: fstrm_tcp_writer_init,
    fstrm_tcp_writer_options_init, fstrm_tcp_writer_options_destroy,
    fstrm_tcp_writer_options_set_socket_address, and
    fstrm_tcp_writer_options_set_socket_port.

  * fstrm_capture: new options for reading from TCP socket.

  * fstrm_capture: add "-c" / "--connections" option to limit the
    number of concurrent connections it will accept.

  * fstrm_capture: add "-b / --buffer-size" option to set the read buffer
    size (effectively the maximum frame size) to a value other than the
    default 256 KiB.

  * fstrm_capture: skip oversize messages to fix stalled connections
    caused by messages larger than the read highwater mark of the input
    buffer. Discarded messages are logged for the purposes of tuning
    the input buffer size.

  * fstrm_capture: complete sending of FINISH frame before closing
    connection.

  * Various test additions and improvements.

fstrm (0.3.2)

  * Accommodate systems without pthread_condattr_setclock (Issue #34)

 -- Chris Mikkelson <cmikk@fsi.io>  Tue, 21 Mar 2017 19:15:11 -0500

fstrm (0.3.1)

  * Add support for '-' as a filename for stdin/stdout (PR #28)

  * destroy condition variable and mutexes in fstrm_iothr_destroy() (PR #25)

 -- Chris Mikkelson <cmikk@fsi.io>  Thu, 26 Jan 2017 16:05:06 -0600

fstrm (0.3.0)

  * Output file rotation (Issue #22).

  * Build on OS X (Issue #21).

  * Change state properly in fstrm_writer_close() (Issue #18).

  * force output flush on sighup (Issue #12).
 
 -- Ben April <bapril@fsi.io>  Fri, 24 Jun 2016 22:29:56 -0500

fstrm (0.2.0)

  * Fix fstrm_writer_open() to allow multiple opens (Issue #1).

  * New utility 'fstrm_capture', which requires libevent >= 2 (Issue #4).

  * Fix incorrect pthread_cond_timedwait() usage (Issue #5).

  * Backwards incompatible library API and protocol changes.

  * Doxygen documentation.

 -- Robert Edmonds <edmonds@fsi.io>  Thu, 06 Nov 2014 19:19:44 -0500

fstrm (0.1.0)

  * Initial release.

 -- Robert Edmonds <edmonds@fsi.io>  Tue, 11 Feb 2014 20:00:25 -0500


src/fstrm_dump.c
/*
 * Copyright (c) 2014, 2018 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <sys/uio.h>
#include <stdio.h>
#include <stdlib.h>

#include <fstrm.h>

#include "libmy/print_string.h"

static fstrm_res
process_start_frame(struct fstrm_reader *r, struct fstrm_writer_options *wopt)
{
	fstrm_res res;
	const struct fstrm_control *control = NULL;
	size_t n_content_type = 0;
	const uint8_t *content_type = NULL;
	size_t len_content_type = 0;

	res = fstrm_reader_get_control(r, FSTRM_CONTROL_START, &control);
	if (res != fstrm_res_success)
		return res;
	fputs("FSTRM_CONTROL_START.\n", stderr);

	res = fstrm_control_get_num_field_content_type(control, &n_content_type);
	if (res != fstrm_res_success)
		return res;
	if (n_content_type > 0) {
		res = fstrm_control_get_field_content_type(control, 0,
			&content_type, &len_content_type);
		if (res != fstrm_res_success)
			return res;
		fprintf(stderr, "FSTRM_CONTROL_FIELD_CONTENT_TYPE (%zd bytes).\n ",
			len_content_type);
		print_string(content_type, len_content_type, stderr);
		fputc('\n', stderr);
	}

	if (wopt != NULL && content_type != NULL) {
		res = fstrm_writer_options_add_content_type(wopt,
			content_type, len_content_type);
		if (res != fstrm_res_success)
			return res;
	}

	return fstrm_res_success;
}

static fstrm_res
print_stop_frame(struct fstrm_reader *r)
{
	fstrm_res res;
	const struct fstrm_control *control = NULL;

	res = fstrm_reader_get_control(r, FSTRM_CONTROL_STOP, &control);
	if (res != fstrm_res_success)
		return res;
	fputs("FSTRM_CONTROL_STOP.\n", stderr);

	return fstrm_res_success;
}

static fstrm_res
print_data_frame(const uint8_t *data, size_t len_data)
{
	fprintf(stderr, "Data frame (%zd) bytes.\n", len_data);
	putchar(' ');
	print_string(data, len_data, stdout);
	putchar('\n');
	return fstrm_res_success;
}

int main(int argc, char **argv)
{
	const char *input_fname = NULL;
	const char *output_fname = NULL;

	fstrm_res res = fstrm_res_failure;
	struct fstrm_file_options *fopt = NULL;
	struct fstrm_writer_options *wopt = NULL;
	struct fstrm_reader *r = NULL;
	struct fstrm_writer *w = NULL;

	int rv = EXIT_FAILURE;

	/* Args. */
	if (argc != 2 && argc != 3) {
		fprintf(stderr, "Usage: %s <INPUT FILE> [<OUTPUT FILE>]\n", argv[0]);
		fprintf(stderr, "Dumps a Frame Streams formatted input file.\n\n");
		return EXIT_FAILURE;
	}
	input_fname = argv[1];
	if (argc == 3)
		output_fname = argv[2];

	/* Line buffering. */
	setvbuf(stdout, NULL, _IOLBF, 0);
	setvbuf(stderr, NULL, _IOLBF, 0);

	/* Setup file reader options. */
	fopt = fstrm_file_options_init();
	fstrm_file_options_set_file_path(fopt, input_fname);

	/* Initialize file reader. */
	r = fstrm_file_reader_init(fopt, NULL);
	if (r == NULL) {
		fputs("Error: fstrm_file_reader_init() failed.\n", stderr);
		goto out;
	}
	res = fstrm_reader_open(r);
	if (res != fstrm_res_success) {
		fputs("Error: fstrm_reader_open() failed.\n", stderr);
		goto out;
	}

	if (output_fname != NULL) {
		/* Setup file writer options. */
		fstrm_file_options_set_file_path(fopt, output_fname);

		/* Setup writer options. */
		wopt = fstrm_writer_options_init();

		/* Copy "content type" from the reader's START frame. */
		res = process_start_frame(r, wopt);
		if (res != fstrm_res_success) {
			fputs("Error: process_start_frame() failed.\n", stderr);
			goto out;
		}

		/* Initialize file writer. */
		w = fstrm_file_writer_init(fopt, wopt);
		if (w == NULL) {
			fputs("Error: fstrm_file_writer_init() failed.\n", stderr);
			goto out;
		}
		res = fstrm_writer_open(w);
		if (res != fstrm_res_success) {
			fstrm_writer_destroy(&w);
			fputs("Error: fstrm_writer_open() failed.\n", stderr);
			goto out;
		}
	} else {
		/* Process the START frame. */
		res = process_start_frame(r, NULL);
		if (res != fstrm_res_success) {
			fprintf(stderr, "Error: process_start_frame() failed.\n");
			goto out;
		}
	}

	/* Loop over data frames. */
	for (;;) {
		const uint8_t *data;
		size_t len_data;

		res = fstrm_reader_read(r, &data, &len_data);
		if (res == fstrm_res_success) {
			/* Got a data frame. */
			res = print_data_frame(data, len_data);
			if (res != fstrm_res_success) {
				fprintf(stderr, "Error: print_data_frame() failed.\n");
				goto out;
			}
			if (w != NULL) {
				/* Write the data frame. */
				res = fstrm_writer_write(w, data, len_data);
				if (res != fstrm_res_success) {
					fprintf(stderr, "Error: write_data_frame() failed.\n");
					goto out;
				}
			}
		} else if (res == fstrm_res_stop) {
			/* Normal end of data stream. */
			res = print_stop_frame(r);
			if (res != fstrm_res_success) {
				fprintf(stderr, "Error: unable to read STOP frame.\n");
				goto out;
			}
			rv = EXIT_SUCCESS;
			break;
		} else {
			/* Abnormal end. */
			fprintf(stderr, "Error: fstrm_reader_read() failed.\n");
			goto out;
		}
	}

out:
	/* Cleanup options. */
	fstrm_file_options_destroy(&fopt);
	fstrm_writer_options_destroy(&wopt);

	/* Cleanup reader. */
	fstrm_reader_destroy(&r);

	/* Cleanup writer. */
	if (w != NULL) {
		res = fstrm_writer_close(w);
		if (res != fstrm_res_success) {
			fprintf(stderr, "Error: fstrm_writer_close() failed.\n");
			rv = EXIT_FAILURE;
		}
		fstrm_writer_destroy(&w);
	}

	return rv;
}
